File: backend/internal/engine/routines.go
package engine

import (
	"botbureau/backend/internal/i18n"

	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"sync"
	"time"
)

// Cap at one year: blocks absurd input and also guards against time overflow
const maxEveryMinutes = 60 * 24 * 365

// Routine is a scheduled task: recurring work dispatched to a bot when it comes due.
type Routine struct {
	Name         string `json:"name"`
	Bot          string `json:"bot"`
	Prompt       string `json:"prompt"`
	EveryMinutes int    `json:"every_minutes"`
	NextRun      int64  `json:"next_run"` // epoch seconds
}

// Scheduler checks every few seconds; each routine that comes due is delivered to its bot as a group-chat message.
type Scheduler struct {
	bus      *Bus
	path     string
	mu       sync.Mutex
	routines map[string]*Routine
	stop     chan struct{}
}

func NewScheduler(bus *Bus, path string) *Scheduler {
	s := &Scheduler{bus: bus, path: path, routines: map[string]*Routine{}, stop: make(chan struct{})}
	s.load()
	return s
}

func clampEvery(v int) int {
	if v < 1 {
		return 1
	}
	if v > maxEveryMinutes {
		return maxEveryMinutes
	}
	return v
}

// ---- Persistence (the file may have been hand-edited: validate each entry, skip bad ones) ----

func (s *Scheduler) load() {
	raw, err := os.ReadFile(s.path)
	if err != nil {
		return
	}
	var items []map[string]any
	if err := json.Unmarshal(raw, &items); err != nil {
		fmt.Fprintf(os.Stderr, i18n.T("%s is not a valid routine list, ignoring its contents\n"), s.path)
		return
	}
	for _, it := range items {
		name, _ := it["name"].(string)
		bot, _ := it["bot"].(string)
		prompt, _ := it["prompt"].(string)
		every, okE := it["every_minutes"].(float64)
		next, okN := it["next_run"].(float64)
		if name == "" || bot == "" || prompt == "" || !okE || !okN {
			fmt.Fprintf(os.Stderr, i18n.T("Skipping a bad entry in routines.json: %v\n"), it)
			continue
		}
		s.routines[name] = &Routine{
			Name: name, Bot: bot, Prompt: prompt,
			EveryMinutes: clampEvery(int(every)), NextRun: int64(next),
		}
	}
}

func (s *Scheduler) save() {
	_ = os.MkdirAll(filepath.Dir(s.path), 0o755)
	out, err := json.MarshalIndent(s.list(), "", "  ")
	if err != nil {
		return
	}
	_ = os.WriteFile(s.path, out, 0o644)
}

func (s *Scheduler) list() []*Routine {
	items := make([]*Routine, 0, len(s.routines))
	for _, r := range s.routines {
		items = append(items, r)
	}
	sort.Slice(items, func(i, j int) bool { return items[i].NextRun < items[j].NextRun })
	return items
}

// ---- Add / remove / list ----

func (s *Scheduler) Add(name, bot, prompt string, everyMinutes int) *Routine {
	everyMinutes = clampEvery(everyMinutes)
	r := &Routine{
		Name: name, Bot: bot, Prompt: prompt,
		EveryMinutes: everyMinutes,
		NextRun:      time.Now().Unix() + int64(everyMinutes)*60,
	}
	s.mu.Lock()
	s.routines[name] = r
	s.save()
	s.mu.Unlock()
	return r
}

func (s *Scheduler) Remove(name string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, ok := s.routines[name]; !ok {
		return false
	}
	delete(s.routines, name)
	s.save()
	return true
}

// RemoveByBot deletes every routine assigned to a bot and returns how many went.

// When someone leaves, the schedules pointing at them should stop. Keeping them waits for a successor
// who can never arrive — ids are one-shot random strings and no new member will ever carry this one
// again — and only drips "routine skipped: its assignee does not exist" into the group forever. Work
// files are what they produced and are kept; a routine is live behaviour and is revoked.
func (s *Scheduler) RemoveByBot(bot string) int {
	s.mu.Lock()
	defer s.mu.Unlock()
	n := 0
	for name, r := range s.routines {
		if r.Bot == bot {
			delete(s.routines, name)
			n++
		}
	}
	if n > 0 {
		s.save()
	}
	return n
}

// Reassign hands a routine to a different bot.

// Until now the only way to "transfer" one was sideways: have the other member save it again under
// the same name and let Add overwrite the entry wholesale (the "same name is overwritten" line in the
// tool description). That asked the user to know an implementation detail and to restate the prompt
// word for word — impossible from a UI that truncates it at 90 characters.
func (s *Scheduler) Reassign(name, bot string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	r, ok := s.routines[name]
	if !ok {
		return false
	}

	// Reassigning does not reset the clock: it changes whose job this is, not when it happens
	r.Bot = bot
	s.save()
	return true
}

func (s *Scheduler) List() []*Routine {
	s.mu.Lock()
	defer s.mu.Unlock()
	out := make([]*Routine, 0, len(s.routines))
	for _, r := range s.routines {
		cp := *r
		out = append(out, &cp)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].NextRun < out[j].NextRun })
	return out
}

// ---- Scheduling loop ----

// tick runs one round: delivers due routines and advances NextRun.
func (s *Scheduler) tick(now int64) {
	var due []*Routine
	s.mu.Lock()
	for _, r := range s.routines {
		if r.NextRun <= now {
			cp := *r
			due = append(due, &cp)

			// Advance arithmetically (rather than accumulating in a loop): this never loops forever regardless of the data,
			// and a backlogged trigger point jumps straight to the next future moment
			step := int64(clampEvery(r.EveryMinutes)) * 60
			r.NextRun += ((now-r.NextRun)/step + 1) * step
		}
	}
	if len(due) > 0 {
		s.save()
	}
	s.mu.Unlock()
	for _, r := range due {
		if s.bus.Deliver("routine:"+r.Name, r.Bot, "group", r.Prompt, true) {
			s.bus.Emit("system", "group", "scheduler",
				fmt.Sprintf(i18n.T("Routine \"%s\" fired, assigned to %s"), r.Name, r.Bot), nil)
		} else {
			s.bus.Emit("system", "group", "scheduler",
				fmt.Sprintf(i18n.T("Routine \"%s\" was skipped: its assignee %s does not exist"), r.Name, r.Bot), nil)
		}
	}
}

func (s *Scheduler) Start() {
	go func() {
		ticker := time.NewTicker(5 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-s.stop:
				return
			case <-ticker.C:
				func() {
					defer func() { // the scheduling loop must never die
						if r := recover(); r != nil {
							s.bus.Emit("system", "", "scheduler",
								fmt.Sprintf(i18n.T("Scheduler error: %v"), r), nil)
						}
					}()
					s.tick(time.Now().Unix())
				}()
			}
		}
	}()
}

File: legacy-python/teambot/routines.py
"""定时例程：bot 学会的重复性任务，按固定间隔自动触发（对应 Grok Bot 的 routines）。"""

from __future__ import annotations

import json
import sys
import threading
import time
from dataclasses import asdict, dataclass

# 间隔上限一年：既挡住荒谬的输入，也保证 next_run 不会溢出 datetime 的表示范围
MAX_EVERY_MINUTES = 60 * 24 * 365


@dataclass
class Routine:
    name: str
    bot: str
    prompt: str
    every_minutes: int
    next_run: float  # epoch 秒


class Scheduler(threading.Thread):
    """后台线程：每隔几秒检查一次，到点的例程作为消息投递给对应 bot。"""

    def __init__(self, bus, path) -> None:
        super().__init__(daemon=True, name="scheduler")
        self.bus = bus
        self.path = path
        self._lock = threading.Lock()
        self._routines: dict[str, Routine] = {}
        self._load()

    # ---- 持久化 ----

    def _load(self) -> None:
        """加载 routines.json。文件可能被手工编辑过，逐条校验，坏条目跳过。"""
        try:
            data = json.loads(self.path.read_text(encoding="utf-8"))
        except (FileNotFoundError, json.JSONDecodeError):
            return
        if not isinstance(data, list):
            print(f"⚠️ {self.path} 顶层不是列表，忽略其中内容", file=sys.stderr)
            return
        for item in data:
            try:
                r = Routine(
                    name=str(item["name"]),
                    bot=str(item["bot"]),
                    prompt=str(item["prompt"]),
                    every_minutes=self._clamp_every(item["every_minutes"]),
                    next_run=float(item["next_run"]),
                )
            except (TypeError, KeyError, ValueError, OverflowError):
                print(f"⚠️ 跳过 routines.json 中的坏条目: {item!r}", file=sys.stderr)
                continue
            self._routines[r.name] = r

    def _save(self) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        payload = [asdict(r) for r in self._routines.values()]
        self.path.write_text(
            json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8"
        )

    @staticmethod
    def _clamp_every(value) -> int:
        return min(max(1, int(value)), MAX_EVERY_MINUTES)

    # ---- 增删查 ----

    def add(self, name: str, bot: str, prompt: str, every_minutes: int) -> Routine:
        every_minutes = self._clamp_every(every_minutes)
        r = Routine(
            name=name,
            bot=bot,
            prompt=prompt,
            every_minutes=every_minutes,
            next_run=time.time() + every_minutes * 60,
        )
        with self._lock:
            self._routines[name] = r
            self._save()
        return r

    def remove(self, name: str) -> bool:
        with self._lock:
            if name not in self._routines:
                return False
            del self._routines[name]
            self._save()
            return True

    def list(self) -> list[Routine]:
        with self._lock:
            return sorted(self._routines.values(), key=lambda r: r.next_run)

    # ---- 调度循环 ----

    def _tick(self, now: float | None = None) -> None:
        """检查一轮：把到点的例程投递出去，并推进 next_run。"""
        now = time.time() if now is None else now
        due: list[Routine] = []
        with self._lock:
            for r in self._routines.values():
                if r.next_run <= now:
                    due.append(r)
                    # 算术推进（而非 while 累加）：无论数据如何都不会死循环，
                    # 且积压的触发点直接跳到下一个未来时刻
                    step = self._clamp_every(r.every_minutes) * 60
                    r.next_run += ((now - r.next_run) // step + 1) * step
            if due:
                self._save()
        for r in due:
            delivered = self.bus.deliver(f"routine:{r.name}", r.bot, r.prompt)
            if delivered:
                self.bus.to_user("scheduler", f"⏰ 例程「{r.name}」触发，已派给 {r.bot}")
            else:
                self.bus.to_user(
                    "scheduler", f"⚠️ 例程「{r.name}」的执行者 {r.bot} 不存在，已跳过"
                )

    def run(self) -> None:
        while True:
            try:
                self._tick()
            except Exception as e:  # 调度线程绝不能死：报告后继续
                self.bus.to_user("scheduler", f"⚠️ 调度器出错：{type(e).__name__}: {e}")
            time.sleep(5)
